rust/src/main.rs

```diff
@@ -45,15 +46,47 @@ fn get_logging_directory() -> Result<PathBuf, io::Error> {
     }
 }
 
-fn path_for_log_file<P: AsRef<Path>>(filename: P) -> Result<PathBuf, io::Error> {
-    let mut logging_directory = get_logging_directory()?;
-    // Create the logging directory
-    fs::create_dir_all(&logging_directory)?;
-    logging_directory.push(filename.as_ref());
-    Ok(logging_directory)
+/// This function tries to create the parent directories for a file
+///
+/// It wraps around the `parent()` function of `Path` which returns an `Option<&Path>` and
+/// `fs::create_dir_all` which returns an `io::Result<()`.
+///
+/// This allows you to use `?`/`try!()` to create the dir and you recive the additional custom error for when `parent()`
+/// returns nothing.
+///
+/// # Errors
+/// This can return an `io::Error` if `fs::create_dir_all` fails or if `parent()` returns `None`.
+/// See `Path`'s `parent()` function for more details.
+/// # Examples
+/// ```
+/// use std::path::PathBuf;
+/// let path_with_file = PathBuf::from("/some/directory/then/file");
+/// assert_eq!(Some(OsStr::new("file")), path_with_file.file_name());
+/// assert_eq!(create_log_directory(path_with_file).is_ok(), true);
+///
+/// let path_with_other_file = PathBuf::from("/other_file");
+/// assert_eq!(Some(OsStr::new("other_file")), path_with_other_file.file_name());
+/// assert_eq!(create_log_directory(path_with_file).is_ok(), true);
+///
+/// // Path that is just the root or prefix:
+/// let path_without_file = PathBuf::from("/");
+/// assert_eq!(None, path_with_file.file_name());
+/// assert_eq!(create_log_directory(path_with_file).is_ok(), false);
+/// ```
+fn create_log_directory(path_with_file: &PathBuf) -> Result<(), io::Error> {
+    let log_dir = path_with_file.parent().ok_or(io::Error::new(
+        io::ErrorKind::InvalidInput,
+        format!(
+            "Unable to get the parent of the following Path: {}",
+            path_with_file.display(),
+        ),
+    ))?;
+    // Try to create the directory.
+    fs::create_dir_all(log_dir)?;
+    Ok(())
 }
 
-fn setup_logging() -> Result<(), fern::InitError> {
+fn setup_logging(logging_path_result: Option<PathBuf>) -> Result<(), fern::InitError> {
```

Comment:
Just `logfile_path` is fine now that it's an Option.

Reply:
Good spot. Done.